lib/persist-state.c

```diff
@@ -267,6 +314,12 @@ _alloc_value(PersistState *self, guint32 orig_size, gboolean in_use, guint8 vers
   persist_state_unmap_entry(self, self->current_ofs);
 
   self->current_ofs += size + sizeof(PersistValueHeader);
+
+  if (!_check_watermark(self) && !_grow_store(self, self->current_size + PERSIST_FILE_INITIAL_SIZE))
+    {
+      main_loop_exit();
+    }
+
   return result;
```

Comment:
On Fri, Nov 13, 2015 at 5:37 PM, juhaszviktor <notifications@github.com>
wrote:

> In lib/persist-state.c
> <https://github.com/balabit/syslog-ng/pull/717#discussion_r44803192>:
>
> > @@ -267,6 +314,12 @@ _alloc_value(PersistState *self, guint32 orig_size, gboolean in_use, guint8 vers
> >    persist_state_unmap_entry(self, self->current_ofs);
> >
> >    self->current_ofs += size + sizeof(PersistValueHeader);
> > +
> > +  if (!_check_watermark(self) && !_grow_store(self, self->current_size + PERSIST_FILE_INITIAL_SIZE))
> > +    {
> > +      main_loop_exit();
> > +    }
> > +
> >    return result;
>
> You are right, we should show an error message about the situation. I will
> add it.
>
> This is a graceful stop, so the destinations send as many messages as they
> can, but as a normal stop there is no guarantee that all the messages are
> sent by the destinations from their queue.
> This patch's goal does not include a better termination, than the normal
> one
>
> I'm not really understand your last note. Almost everything is managed by
> the main_loop, as the control socket and debugger, even so, they call the
> main_loop exit.
>
Both the control socket and the debugger is pretty tightly coupled with the
mainloop, but you are right this coupling should be lessened. persist-state
however was so far completely independent and the only reason we want to
introduce this coupling is because of a global error condition, that could
be signaled by a simple callback.